### PR TITLE
TextHighlight: Convert component to TypeScript

### PR DESCRIPTION
**Pull request by walbo**
## What?
Converts the `TextHighlight` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (https://github.com/WordPress/gutenberg/issues/35744).

## How?
- Refactors the current `TextHighlight` to TypeScript.
- Refactors the test to use `@testing-library/react'`
- Refactors the storybook story

## Testing Instructions
1. Run npm run dev and ensure no typing errors
2. Fire up the Storybook and confirm the `TextHighlight` continues to function as expected
3. Test the various TextHighlight within the block/site editors in Gutenberg still work

### CODE REVIEW

File: packages/components/src/text-highlight/index.tsx
Comment (walbo) [inline packages/components/src/text-highlight/index.tsx:36]: Wrapped the text in a fragment to always return a `JSX.Element`. TypeScript was complaining otherwise. Not sure if there is another way, but this shouldn't break anything.

File: packages/components/src/text-highlight/test/index.tsx
Comment (mirka) [inline packages/components/src/text-highlight/test/index.tsx:14]: TIL the `mark` role is [not officially in the spec](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/mark_role#specifications) yet, thus not supported by `getByRole`!

File: packages/components/src/text-highlight/test/index.tsx
Comment (mirka) [inline packages/components/src/text-highlight/test/index.tsx]: Just curious: Have you considered using [`.toHaveTextContent()`](https://github.com/testing-library/jest-dom#tohavetextcontent) for these? I find it a bit more ergonomic and less error-prone, but maybe it's just personal preference.

```suggestion
					expect( el ).toHaveTextContent(
						new RegExp( `^${ highlight }$` )
					);
```

**Review by mirka — APPROVED** [review]
Looks good! The `.toHaveTextContent()` thing is not a blocker if you prefer the current approach 🙂

File: packages/components/src/text-highlight/index.tsx
Comment (ciampo) [inline packages/components/src/text-highlight/index.tsx:32]: We should probably either remove the fallback `''` value (since `text` and `highlight` are marked as required props), or consider marking them as optional and keep the feedback value (although I believe that at least `text` should be required).

Although I understand that we may want to keep this for backwards compat to be extra sure that these props are always defined?

Also, in case we keep the default value, we may want to add that to the JSDocs (via `@default` notation) and to the README.

File: packages/components/src/text-highlight/index.tsx
Comment (ciampo) [inline packages/components/src/text-highlight/index.tsx:31]: This is an interesting example of how ref forwarding and using the `WordPressComponentProps` type don't make much sense with the current way the component is written (cc @mirka )

File: packages/components/src/text-highlight/test/index.tsx
Comment (ciampo) [reply packages/components/src/text-highlight/test/index.tsx:14]: TIL too! Should we add a comment about this, saying that the code should be updated to use its intrinsic `role` once the spec update lands?

File: packages/components/src/text-highlight/test/index.tsx
Comment (ciampo) [inline packages/components/src/text-highlight/test/index.tsx]: I know this comment existed prior to this PR, but I wonder if we could rewrite it as it isn't very clear to me.

File: packages/components/src/text-highlight/test/index.tsx
Comment (ciampo) [inline packages/components/src/text-highlight/test/index.tsx]: Same about this comment

File: packages/components/src/text-highlight/index.tsx
Comment (walbo) [reply packages/components/src/text-highlight/index.tsx:32]: I think it makes sence to have them both marked as required. Not sure why you want to use this component if not both is provided?

Not sure I want to change the behavor in this PR. Trying to not change the behavor when refactoring to TypeScript. Maybe do a follow-up PR and it can be discussed more there?

Have added the `@default` notation.

File: packages/components/src/text-highlight/test/index.tsx
Comment (walbo) [reply packages/components/src/text-highlight/test/index.tsx:14]: Added a comment about it.

File: packages/components/src/text-highlight/test/index.tsx
Comment (walbo) [reply packages/components/src/text-highlight/test/index.tsx]: Updated the comment.

File: packages/components/src/text-highlight/test/index.tsx
Comment (walbo) [reply packages/components/src/text-highlight/test/index.tsx]: Updated.

**Comment by walbo** [comment]
Anything else I need to address here? Would love to get a final review after the last commits.

**Comment by ciampo** [comment]
> Anything else I need to address here? Would love to get a final review after the last commits.

Hey @walbo ! Lena is currently AFK, and I'm catching up with all the PR review requests that I accumulated during the past week while I was AFK. I'll try to get to this PR (and the others that you have open) as quickly as possible! Thank you for your patience 🙏

File: packages/components/src/text-highlight/index.tsx
Comment (ciampo) [reply packages/components/src/text-highlight/index.tsx:32]: > Not sure I want to change the behavor in this PR. Trying to not change the behavor when refactoring to TypeScript.

Good point, sounds sensible

**Review by ciampo — APPROVED** [review]
LGTM 🚀  Thank you!

**Comment by walbo** [comment]
> > Anything else I need to address here? Would love to get a final review after the last commits.
> 
> Hey @walbo ! Lena is currently AFK, and I'm catching up with all the PR review requests that I accumulated during the past week while I was AFK. I'll try to get to this PR (and the others that you have open) as quickly as possible! Thank you for your patience 🙏

Thanks for the review! No worries. No rush for me. ;)

**Pull request merged**
